build-system/compile/json-schema/index.js

```diff
@@ -0,0 +1,121 @@
+  md5,
+} = require('../../common/transform-cache');
+
+/**
+ * File filter to match import.
+ * It's useful that this always ends in .schema.json, since that suffix enables
+ * JSON Schema features on IDEs like VSCode.
+ * @type {RegExp}
+ */
+const importFilter = /\.schema\.json$/;
+
+// https://ajv.js.org/options.html
+const ajvOptions = {
+  code: {
+    esm: true,
+    source: true,
+    lines: true,
+  },
+  allErrors: true,
+  validateSchema: false,
+};
+
+const ajv = once(() => new Ajv(ajvOptions));
+
+const escapeJsIdentifier = (id) => id.replace(/[^a-z_0-9]/gi, '_');
+
+const getTransformCache = once(() => new TransformCache('.json-schema-cache'));
+
+/**
+ * Like batchRead(), but considers ajvOptions part of the invalidation hash.
+ * @param {string} filename
+ * @return {Promise<{contents: string, hash: string}>}
+ */
+async function ajvBatchedRead(filename) {
+  const {contents, hash} = await batchedRead(filename);
+  const rehash = md5(JSON.stringify({ajvOptions, hash}));
+  return {contents, hash: rehash};
+}
+
+/**
+ * @param {string} filename
+ * @return {Promise<string>}
+ */
+async function getCompiledJsonSchemaFilename(filename) {
+  const {contents, hash} = await ajvBatchedRead(filename);
+  const transformCache = getTransformCache();
+  const cached = await transformCache.get(hash);
+  let output = cached;
+  if (!output) {
+    output = avjCompile(filename, contents);
+    transformCache.set(hash, Promise.resolve(output));
+  }
+  const outputFilename = `build/${filename}.js`;
+  if (!cached || !(await pathExists(outputFilename))) {
+    await outputFile(outputFilename, output);
+  }
+  return outputFilename;
+}
+
+/**
+ * @param {string} filename
+ * @param {string} contents
+ * @return {string}
+ */
+function avjCompile(filename, contents) {
+  const schema = JSON.parse(contents);
+  const validateAjv = ajv().compile(schema);
+  const validateFnName = escapeJsIdentifier(
+    `validate_${basename(filename, '.json')}`
+  );
+  return (
+    dedent(`
+      const validateAjv = __validateAjv__;
+      /**
+       * @param {any} data
+       * @return {{params?: object, message?: string, data?: any}[]}
+       * https://ajv.js.org/api.html#error-objects
+       */
+      export default function __validateFnName__(data) {
+        return validateAjv(data) ? [] : [...validateAjv.errors]
+      }
+    `)
+      .replace('__validateFnName__', validateFnName)
+      // validateAjv returns a boolean, and modifies a property of the function
+      // for errors. The default instead returns an array of errors, which is
+      // empty when the input is valid.
+      .replace('__validateAjv__', validateAjv.toString())
+  );
+}
+
+const esbuildJsonSchemaPlugin = {
+  name: 'json-schema',
+  setup(build) {
+    build.onResolve({filter: importFilter}, async ({path, resolveDir}) => {
+      const filename = relative(process.cwd(), join(resolveDir, path));
+      const compiledFilename = await getCompiledJsonSchemaFilename(filename);
```

Comment:
This will force the plugin to compile (or attempt to) every time it's included, even if it's unused. We should split this into a `build.onLoad` phase, which can then do the compilation as it's actually needed. Eg, set the `namespace` here to `ajv`, then capture any `ajv` loads to perform the compilation.

Reply:
I tried this, but it makes our build pipeline fail:

```
ERROR: Error: Build failed with 1 error:
/Users/alanorozco/git/amphtml/build-system/common/esbuild-babel.js:155:20: error: [plugin: babel] failed to find sourcemap for babel file "json-schema:build-system/compile/json-schema/test/input/my-schema.schema.json"
    at failureErrorWithLog (/Users/alanorozco/git/amphtml/node_modules/esbuild/lib/main.js:1493:15)
    at /Users/alanorozco/git/amphtml/node_modules/esbuild/lib/main.js:1151:28
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
```

esbuild plugin order does not seem to change this.

Reply:
Ok so now I tried this:
```js
  setup(build) {
    build.onLoad({filter: importFilter, namespace: ''}, async ({path}) => ({
      contents: await ajvCompileCached(path),
    }));
  }
```
And I get almost the same error, except the path reported is different:
```
[17:31:43] ERROR: Error: Build failed with 1 error:
/Users/alanorozco/git/amphtml/build-system/common/esbuild-babel.js:157:20: error: [plugin: babel] failed to find sourcemap for babel file "../../../../../../../Users/alanorozco/git/amphtml/build-system/compile/json-schema/test/input/my-schema.schema.json"
```